### DISCUSSION TITLE
Custom Azure clouds in Grafana

### DISCUSSION THREAD

**Discussion by kostrse** [grafana/grafana#61016] [Data Sources]
Each Azure datasource in Grafana has a predefined list of supported Azure clouds: Public, China and US Government.

Currently, Its impossible to use the existing Azure datasources for Azure Stack or other hybrid cloud scenarios because it’s not possible to configure a custom cloud environment.

The proposal is to address this limitation by introducing cloud configuration via Grafana settings and [Grafana Azure SDK](https://github.com/grafana/grafana-azure-sdk-go).

This would allow to uniformly support custom clouds across all Azure datasources.

### Grafana config
First step would be to introduce a new optional property to the `[azure]`  section of Grafana config, `cloud_config`:

```ini
[azure]
cloud = AzureCloud
...
cloud_config = ./clouds.json
```

This property should point to a JSON file with definitions of all available clouds. If this property isn’t set, then the built-in list of clouds should be initialized (Public, China, US Gov), same as the existing behavior.

The JSON file would have the same structure as returned by Azure CLI:
```
az cloud list --output json > ./clouds.json
```

Example of such cloud config (returned by `az cloud list`):
```json
[
  {
    "name": "AzureCloud",
    "isActive": true,
    "profile": "latest",
    "endpoints": {
      "activeDirectory", "https://login.microsoftonline.com",
      "portal", "https://portal.azure.com",
      "...": "..."
    },
    "suffixes": {
      "...": "..."
    },
  },
  "...": "..."
]
```

It has a `name` property which is a cloud identifier, same us currently used for built-in clouds ([here](https://github.com/grafana/grafana-azure-sdk-go/blob/3ba629ee4f48e9ded325bb3d5e34bf5649e79237/azsettings/clouds.go#L6)) and two key/value bags `endpoints ` and `suffixes`. Properties `isActive` and `profile` are not relevant for the current use case.

Path to the clouds file should be relative to the location of the `grafana.ini` config (`$WORKING_DIR/conf`).

### New API in Grafana Azure SDK

Here are new structures and functions proposed for the Grafana Azure SDK:

```go
type azcloud.AzureCloud struct {
	Name        string
	DisplayName string
}

type azcloud.AzureCloudDef struct {
	Name        string
	DisplayName string
	Endpoints   map[string]string
	Suffixes    map[string]string
}

func azcloud.InitDefaultClouds() error
func azcloud.InitFromConfig(configFile string) error

func azcloud.GetClouds() ([]AzureCloud, error)

func azcloud.GetCloud(azureCloud string) (*AzureCloudDef, error)
```

First, clouds configuration has to be initialized at startup using either `InitDefaultClouds() ` or `InitFromConfig(configFile)` . It needs to be called by Grafana host during initialization of settings ([here](https://github.com/grafana/grafana/blob/12c292fd446f855d2a42a3c504b59edb6f51ac2c/pkg/setting/setting_azure.go#L5)) or by external plugin in its main function.

Once initialized, the list of clouds stored in memory as a singleton object and available via the following API:
`GetClouds()`  returns a list of all available clouds and `GetCloud(azureCloud)`  returns configuration of the given cloud.

These functions can be called by business logic of datasources to resolve cloud-specific resource endpoints and also by Grafana Azure SDK itself to resolve the AAD endpoint for authentication (e.g. [here](https://github.com/grafana/grafana-azure-sdk-go/blob/3ba629ee4f48e9ded325bb3d5e34bf5649e79237/aztokenprovider/token_provider.go#L115)).

### List of available clouds in frontend UI

Since actual authentication performed on backend, frontend only needs a list of cloud names to display in datasource configuration UI.

Grafana host can pass the list of clouds to frontend via the `frontendsettings` ([here](https://github.com/grafana/grafana/blob/9bd6e471e436fed37a908e5118593d515b134781/pkg/api/frontendsettings.go#L184)):

```go
jsonObj := map[string]interface{}{
	...
	"azure": map[string]interface{}{
		"cloud":  hs.Cfg.Azure.Cloud,
		...
		"clouds": azcloud.GetCouds(),
	},
}
```

Frontend doesn’t need all cloud configuration properties, only cloud names.

### Access to clouds configuration from external plugin backend

Grafana passes configuration from the host to external plugins via environment variables ([here](https://github.com/grafana/grafana/blob/4d8287b319514b750617c20c130ffc424a3ecf2c/pkg/plugins/manager/loader/initializer/initializer.go#L71)).
Clouds config is a large structure, it could be problematic to pass it via environment variables.
Currently, it’s not clear what is the best solution, but one would be to pass absolute path to the JSON file, so external plugin could read and parse it by itself.

### Missing endpoints and extensibility of the clouds config JSON

The idea to reuse the existing format of clouds config from `az cloud list` is to avoid inventing own format and make it simpler to write deployment scripts which could reuse naming of properties etc.

It’s possible that the original cloud config from `az cloud list` doesn’t contain all needed properties, so it can be extended as needed, while reusing existing properties which already available.

Here are some properties which are missing in the original config and would need to be introduced:

| Property | Description |
| - | - |
| displayName | The name of the cloud to display in the UI, as opposed to _name_ which is more like an identifier rather than a user-friendly name. |
| _TBD_ |  |

**Comment by aangelisc** [comment]
Thanks for opening this discussion @kostrse! This is a great plan for implementation of this feature.

Some questions from me:

With regards to initialising the default clouds, can this be done via API rather than being hardcoded? It seems the `az cloud list` command doesn't require an account to be logged in, is there an equivalent REST API that could be used to retrieve default clouds? Would this also work for retrieving custom clouds running on Azure Stack or does this require a user to be logged in?

Would the usage of Grafana on Azure Stack/hybrid cloud introduce any other substantial differences in the datasources or do the currently supported services function identical to public cloud?

Is the `suffixes` property needed? It currently isn't used in either datasource.

> Currently, it’s not clear what is the best solution, but one would be to pass absolute path to the JSON file, so external plugin could read and parse it by itself.

If we were to use the file path passed in config solution then I agree that the external plugin should read and parse the file.

On the missing `displayName`, I guess we could infer this from the `name` property?

**Reply by kostrse** [reply]
> With regards to initialising the default clouds, can this be done via API rather than being hardcoded? It seems the `az cloud list` command doesn't require an account to be logged in, is there an equivalent REST API that could be used to retrieve default clouds? Would this also work for retrieving custom clouds running on Azure Stack or does this require a user to be logged in?

There's no such API. Azure CLI comes with a default list of clouds but the list is stored locally and can be customized by user.

- https://learn.microsoft.com/en-us/cli/azure/manage-clouds-azure-cli#register-a-new-cloud
- https://learn.microsoft.com/en-us/cli/azure/cloud?view=azure-cli-latest#az-cloud-register-examples

Another thing to consider is that this initialization is going to run during startup of Grafana regardless user needs it or not. Initialization of a hardcoded structure has zero cost, dependency on external resource would be blocking the startup.

The list of default clouds and endpoints is already hardcoded in several places, in each datasource, this change would make it hardcoded in a single place - [github.com/grafana/grafana-azure-sdk-go](https://github.com/grafana/grafana-azure-sdk-go).

We don't even need all properties to be included in the built-in list, only ones which actually used by Azure datasources. So bult-in structure wouldn't be large.

> Would the usage of Grafana on Azure Stack/hybrid cloud introduce any other substantial differences in the datasources or do the currently supported services function identical to public cloud?

By design there should not be any substantial differences. All clouds supposed to behave identically. There are exceptions, not always all services or features available in all environments, but nothing should require more than conditional logic based on feature checks (e.g. if endpoint is set to `null` then hide feature).

> Is the `suffixes` property needed? It currently isn't used in either datasource.

I would include it for sake of completeness.
For built-in list of default clouds, we don't need to populate it if we aren't going to use any properties from it for now.

I see it would be useful for whitelisting/validation of endpoints entered by user.
For example, it could be used to construct authentication scopes in Azure Data Explorer [here](https://github.com/grafana/azure-data-explorer-datasource/blob/9a1b3b166a270e8256463e971f0619a12536ad1b/pkg/azuredx/adxauth/scopes.go#L13) based on account name rather than blindly using `clusterUrl` as provided by user.

> On the missing `displayName`, I guess we could infer this from the `name` property?

It can be optional property, used if present.

The cloud `name` is an identifier, stored in JsonData, it can be `AzureCloud` or can be `eus2-4567-kihynvxmku`

It would be used to populate this list:
- https://github.com/grafana/grafana/blob/122be161f9af2d3c4a51aadc7d94f30a3d9650e2/public/app/plugins/datasource/prometheus/configuration/AzureCredentials.ts#L10

**Comment by andresmgot** [comment]
> ### Access to clouds configuration from external plugin backend
> Grafana passes configuration from the host to external plugins via environment variables ([here](https://github.com/grafana/grafana/blob/4d8287b319514b750617c20c130ffc424a3ecf2c/pkg/plugins/manager/loader/initializer/initializer.go#L71)). Clouds config is a large structure, it could be problematic to pass it via environment variables. Currently, it’s not clear what is the best solution, but one would be to pass absolute path to the JSON file, so external plugin could read and parse it by itself.

I think having this configuration as a different file could be problematic. In some environments, it may not be possible (or difficult) to attach external files. Also, in the future, external plugins may even run in a different host than Grafana so plugins won't have direct access to files.

Because of that, I think it would be better to dump the JSON object in the property:

```ini
[azure]
cloud = AzureCloud
...
cloud_config = `[
  {
    "name": "AzureCloud",
    "isActive": true,
    "profile": "latest",
    "endpoints": {
      "activeDirectory", "https://login.microsoftonline.com",
      "portal", "https://portal.azure.com",
      "...": "..."
    },
    "suffixes": {
      "...": "..."
    },
  },
  "...": "..."
]`
```

There is another example like this in the current [configuration](https://github.com/grafana/grafana/pull/37069/files).

I see your comment about the length of this configuration. Testing this locally, my config is ~8k characters long while the maximum size for an environment variable is ~32k which should be enough. In any case, if some limit is hit, the file can be manually edited to remove unused information.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82915. Feel free to continue the discussion around this there. Thank you!
